### PR TITLE
Canvas: Add corner radius option

### LINKED ISSUE #77526: Canvas: Rectangle Element rounded corner option
**Why is this needed**:

Better display of rectangle elements

**What would you like to be added**:

Configuration of the _rounded_ parameter for rectangle elements in the Element configuration section.
Permit a number slider for the amount of rounding 0 to X.

**Who is this feature for?**

When creating status panels rounded rectangles and such are much easier to look at for hours on end.

### PR DESCRIPTION
**Pull request by drew08t**
Adds option for corner radius for elements and connections.

- [x] Calcs for connection vertex radius
- [x] Connection radius option
- [x] Element border radius option
- [x] Constraints on connection radius (if curve < half-segment, reduce radius)
- [x] Check geometry for tangency
- [ ] ~Clean up and simplify math~ https://github.com/grafana/grafana/issues/85013
- [x] Rebase to include all new elements (https://github.com/grafana/grafana/pull/84205)

Mar-22-2024 00-40-27

Closes #77526

### CODE REVIEW

**Comment by drew08t** [comment]
The reason it is showing up "bumpy" for connections is because the scaling in x and y are different from vertex relative coordinates to canvas absolute coordinates. To fix this, should explore quadratic bezier at corners, instead of arcs.

**Comment by drew08t** [comment]
> The reason it is showing up "bumpy" for connections is because the scaling in x and y are different from vertex relative coordinates to canvas absolute coordinates. To fix this, should explore quadratic bezier at corners, instead of arcs.

Actually quadratic bezier is ideal because of constraints limiting the radius and throwing off tangency; however, the real fix was to do tangency calculations in the scene coordinate system versus the possibly skewed vertex coordinate system.

File: public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx [hunk 1]
```diff
@@ -151,26 +152,112 @@ export const ConnectionSVG = ({
...
       // Create vertex path and populate array of add vertex controls
       const addVertices: ConnectionCoordinates[] = [];
       let pathString = `M${x1} ${y1} `;
       if (vertices?.length) {
         vertices.map((vertex, index) => {
           const x = vertex.x;
           const y = vertex.y;
-          pathString += `L${x * (x2 - x1) + x1} ${y * (y2 - y1) + y1} `;
+
+          const X = x * (x2 - x1) + x1;
```
Comment (drew08t) [inline public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx:164]: This whole section can be significantly simplified and made more robust to properly handle clockwise vs counter clockwise combinations.

File: public/app/plugins/panel/canvas/utils.ts [hunk 2]
```diff
@@ -371,7 +371,12 @@ export const calculateAbsoluteCoords = (
 };
 
 export const calculateAngle = (x1: number, y1: number, x2: number, y2: number) => {
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/utils.ts:373]: Minor add what this is taking / returns (minor inline docs)

File: public/app/plugins/panel/canvas/editor/options.ts [hunk 3]
```diff
@@ -129,6 +141,27 @@ export const optionBuilder: OptionSuppliers = {
...
   },
 
+  addRadius: (builder, context) => {
+    const category = ['Radius'];
+    builder.addCustomEditor({
+      category,
+      id: 'radius',
+      path: 'radius',
+      name: 'Radius',
+      editor: ScaleDimensionEditor,
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/editor/options.ts:151]: Only allow selection of numeric based fields 
Screenshot 2024-03-22 at 11 20 49 AM

File: public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx [hunk 4]
```diff
@@ -151,27 +155,163 @@ export const ConnectionSVG = ({
...
+          // Add segment to path
+          pathString += `L${xa} ${ya} `;
+
+          if (lHalfArc !== 0) {
+            // Add arc if applicable
+            pathString += `Q ${X} ${Y} ${xb} ${yb} `;
+          }
         });
+        // Add last segment
         pathString += `L${x2} ${y2}`;
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx:331]: minor for later but ability to draw vertex over actual line / some way of indicating relation or something
Screenshot 2024-03-22 at 11 23 40 AM

File: public/app/features/canvas/runtime/element.tsx [hunk 5]
```diff
@@ -358,6 +358,10 @@ export class ElementState implements LayerElement {
       }
     }
 
+    if (border && border.radius !== undefined) {
+      css.borderRadius = `${border.radius}px`;
+    }
```
Comment (nmarrs) [inline public/app/features/canvas/runtime/element.tsx:363]: For svg based elements (i.e. ellipse, triangle, parellogram, and cloud?) we need to apply border radius to actual svg (similar to background image etc)

TODO create a follow-up GH issue for this

Screenshot 2024-03-22 at 11 33 11 AM

File: public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx [hunk 6]
```diff
@@ -151,27 +155,163 @@ export const ConnectionSVG = ({
...
       // Create vertex path and populate array of add vertex controls
       const addVertices: ConnectionCoordinates[] = [];
       let pathString = `M${x1} ${y1} `;
       if (vertices?.length) {
         vertices.map((vertex, index) => {
           const x = vertex.x;
           const y = vertex.y;
-          pathString += `L${x * (x2 - x1) + x1} ${y * (y2 - y1) + y1} `;
+
+          // Convert vertex relative coorindates to scene coordinates
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx:167]:
```suggestion
          // Convert vertex relative coordinates to scene coordinates
```

File: public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx [hunk 7]
```diff
@@ -151,27 +155,163 @@ export const ConnectionSVG = ({
...
+          let xb = X;
+          let yb = Y;
+
+          // Initialize half arc distance and segment angles
+          let lHalfArc = 0;
+          let angle1 = 0;
+          let angle2 = 0;
+
+          // Only calculate arcs if there is a radius
+          if (radius) {
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx:185]: maybe add check that radius && radius != 0?

File: public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx [hunk 8]
```diff
@@ -151,27 +155,163 @@ export const ConnectionSVG = ({
...
+              lHalfArc *= -1;
+            }
+          }
+
           if (index === 0) {
             // For first vertex
             addVertices.push(calculateMidpoint(0, 0, x, y));
+
+            // Only calculate arcs if there is a radius
+            if (radius) {
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx:230]: similar to check above?

File: public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx [hunk 9]
```diff
@@ -151,27 +155,163 @@ export const ConnectionSVG = ({
...
+                yb = -lHalfArc * Math.sin(angle2) + Y;
+              }
+            }
           } else {
             // For all other vertices
             const previousVertex = vertices[index - 1];
             addVertices.push(calculateMidpoint(previousVertex.x, previousVertex.y, x, y));
+
+            // Only calculate arcs if there is a radius
+            if (radius) {
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/components/connections/ConnectionSVG.tsx:274]: similar check above

**Review by nmarrs — APPROVED** [review]
♾️

**Review by Develer — APPROVED** [review]

File: public/app/plugins/panel/canvas/editor/options.ts (hunk 3 above)
Comment (drew08t) [reply public/app/plugins/panel/canvas/editor/options.ts:151]: Created a follow-up issue for this here: https://github.com/grafana/grafana/issues/85019

**Review by codeincarnate — APPROVED** [review]
Tested this out and it looks good!

**Pull request merged**